contracts/badger-sett/strategies/unitprotocol/StrategyUnitProtocolMeta.sol

```diff
@@ -0,0 +1,248 @@
+            debt = debtAmt.sub(maxDebt);
+        }
+
+        return (debt > 0) ? debt : 0;
+    }
+
+    function _getBufferedMinRatio(uint256 _multiplier) internal view returns (uint256) {
+        require(ratioBuffMax > 0, "!ratioBufferMax");
+        require(minRatio > 0, "!minRatio");
+        return minRatio.mul(_multiplier).mul(ratioBuffMax.add(ratioBuff)).div(ratioBuffMax).div(100);
+    }
+
+    function borrowableDebt() public view returns (uint256) {
+        uint256 collateralAmt = getCollateralBalance();
+        return calculateDebtFor(collateralAmt, true);
+    }
+
+    function requiredPaidDebt(uint256 _redeemCollateralAmt) public view returns (uint256) {
+        uint256 collateralAmt = getCollateralBalance().sub(_redeemCollateralAmt);
+        return calculateDebtFor(collateralAmt, false);
+    }
+
+    // **** sub-strategy implementation ****
+    function _depositUSDP(uint256 _usdpAmt) internal virtual;
+
+    function _withdrawUSDP(uint256 _usdpAmt) internal virtual;
+
+    // **** Oracle (using chainlink) ****
+
+    function getLatestCollateralPrice() public view returns (uint256) {
+        require(unitOracle != address(0), "!_collateralOracle");
+
+        (, int256 price, , , ) = IChainlinkAggregator(unitOracle).latestRoundData();
+
+        if (price > 0) {
+            int256 ethPrice = 1;
+            if (collateralPriceEth) {
+                (, ethPrice, , , ) = IChainlinkAggregator(eth_usd).latestRoundData(); // eth price from chainlink in 1e8 decimal
+            }
+            return uint256(price).mul(collateralPriceDecimal).mul(uint256(ethPrice)).div(1e8).div(collateralPriceEth ? 1e18 : 1);
+        } else {
+            return 0;
+        }
+    }
+
+    // **** Setters ****
+
+    function setMinRatio(uint256 _minRatio) external {
+        _onlyGovernance();
+        minRatio = _minRatio;
+    }
+
+    function setRatioBuff(uint256 _ratioBuff) external {
+        _onlyGovernance();
+        ratioBuff = _ratioBuff;
+    }
+
+    // **** Unit Protocol CDP actions ****
+
+    function addCollateralAndBorrow(uint256 _collateralAmt, uint256 _usdpAmt) internal {
+        require(_usdpAmt.add(debtUsed()) < debtLimit(), "!exceedLimit");
+        _safeApproveHelper(collateral, unitVault, _collateralAmt);
+        IUnitCDPManager(cdpMgr01).join(collateral, _collateralAmt, _usdpAmt);
+    }
+
+    function repayAndRedeemCollateral(uint256 _collateralAmt, uint256 _usdpAmt) internal {
+        _safeApproveHelper(debtToken, unitVault, _usdpAmt);
+        IUnitCDPManager(cdpMgr01).exit(collateral, _collateralAmt, _usdpAmt);
+    }
+
+    // **** State Mutation functions ****
+
+    function keepMinRatio() external {
+        _onlyCDPInUse();
+        _onlyAuthorizedActorsOrController();
+
+        uint256 requiredPaidback = requiredPaidDebt(0);
+        if (requiredPaidback > 0) {
+            _withdrawUSDP(requiredPaidback);
+
+            uint256 _actualPaidDebt = IERC20Upgradeable(debtToken).balanceOf(address(this));
```

Comment:
To avoid wasting gas by calling `IERC20Upgradeable(debtToken).balanceOf(address(this))` again on line 186, which will return the same value, we can store it in a different variable:

```solidity
uint256 _debtTokenBalance = IERC20Upgradeable(debtToken).balanceOf(address(this));
uint256 _fee = getDebtBalance().sub(getDebtWithoutFee());

require(_debtTokenBalance > _fee, "!notEnoughForFee");
uint256 _actualPaidDebt = _debtTokenBalance.sub(_fee); // unit protocol will charge fee first
 _actualPaidDebt = _capMaxDebtPaid(_actualPaidDebt);

require(_debtTokenBalance >= _actualPaidDebt.add(_fee), "!notEnoughRepayment");
```